src/impl_methods.rs

```diff
@@ -228,25 +235,117 @@ impl<A, S, D> ArrayBase<S, D> where S: Data<Elem=A>, D: Dimension
-    pub fn slice_mut(&mut self, indexes: &D::SliceArg) -> ArrayViewMut<A, D>
-        where S: DataMut
+    /// (**Panics** if `D` is `IxDyn` and `info` does not match the number of array axes.)
+    pub fn slice_mut<I, T, Do>(&mut self, info: I) -> ArrayViewMut<A, Do>
+    where
+        I: Borrow<SliceInfo<T, Do>>,
+        T: Borrow<D::SliceArg>,
+        Do: Dimension,
+        S: DataMut,
     {
-        let mut arr = self.view_mut();
-        arr.islice(indexes);
-        arr
+        self.view_mut().slice_into(info)
     }
 
-    /// Slice the array’s view in place.
+    /// Slice the array’s view, possibly changing the number of dimensions.
     ///
     /// See also [`D::SliceArg`].
     ///
     /// [`D::SliceArg`]: trait.Dimension.html#associatedtype.SliceArg
     ///
     /// **Panics** if an index is out of bounds or stride is zero.<br>
-    /// (**Panics** if `D` is `IxDyn` and `indexes` does not match the number of array axes.)
-    pub fn islice(&mut self, indexes: &D::SliceArg) {
-        let offset = D::do_slices(&mut self.dim, &mut self.strides, indexes);
+    /// (**Panics** if `D` is `IxDyn` and `info` does not match the number of array axes.)
+    pub fn slice_into<I, T, Do>(mut self, info: I) -> ArrayBase<S, Do>
+    where
+        I: Borrow<SliceInfo<T, Do>>,
+        T: Borrow<D::SliceArg>,
+        Do: Dimension,
+    {
+        let info: &SliceInfo<_, _> = info.borrow();
+        let indices: &[SliceOrIndex] = info.indices().borrow().borrow();
+
+        // Slice and subview in-place without changing the number of dimensions.
+        self.islice(indices);
+
+        // Copy the dim and strides that remain after removing the subview axes.
+        let mut new_dim = Do::zero_index_with_ndim(info.out_ndim());
+        let mut new_strides = Do::zero_index_with_ndim(info.out_ndim());
+        let remaining = izip!(self.dim.slice(), self.strides.slice(), indices).filter_map(
+            |(d, s, slice_or_index)| match slice_or_index {
+                &SliceOrIndex::Slice(_) => Some((d, s)),
+                &SliceOrIndex::Index(_) => None,
+            },
+        );
+        for ((d, s), (new_d, new_s)) in
+            remaining.zip(izip!(new_dim.slice_mut(), new_strides.slice_mut()))
+        {
+            *new_d = *d;
+            *new_s = *s;
+        }
+
+        ArrayBase {
+            ptr: self.ptr,
+            data: self.data,
+            dim: new_dim,
+            strides: new_strides,
+        }
+    }
+
+    /// Slice the array’s view in place without changing the number of dimensions.
+    ///
+    /// See also [`D::SliceArg`].
+    ///
+    /// [`D::SliceArg`]: trait.Dimension.html#associatedtype.SliceArg
+    ///
+    /// **Panics** if an index is out of bounds or stride is zero.<br>
+    /// (**Panics** if `D` is `IxDyn` and `info` does not match the number of array axes.)
+    pub fn islice<I>(&mut self, info: I)
+    where
+        I: AsRef<[SliceOrIndex]>,
```

Comment:
This method loses the static length check this way. It would be ideal to keep it, but it's a less used method so not that important.

Reply:
Good catch. I just pushed a fix to keep compile-type length checking.